### DISCUSSION TITLE
Downscaling issue with QR codes

### DISCUSSION THREAD

**Discussion by natehitze** [SixLabors/ImageSharp#1780] [Q&A] [answered]
I'm working on a generic downscaling process for user image uploads, using this setup (this is from a POC project):

```csharp
originalImage.Mutate(x => x
    .BackgroundColor(new Rgba32(0, 0, 0, 0))
    .Resize(new ResizeOptions()
    {
        Size = new Size(200, 200),
        Mode = SixLabors.ImageSharp.Processing.ResizeMode.Max,
        Sampler = KnownResamplers.Lanczos5,
        Compand = true,
    })
    .GaussianSharpen(0.3f));
    var processedStream = new MemoryStream();
    originalImage.Save(processedStream, new PngEncoder()
    {
        CompressionLevel = PngCompressionLevel.BestCompression,
        TransparentColorMode = PngTransparentColorMode.Clear,
    });
```
I get good results with this for most images, including images that are just text. The only image I'm having a problem with is a QR code. Here's the before and after:

![image](https://user-images.githubusercontent.com/3506334/137364627-d12a604e-f98f-42fd-83e9-ae4de33273b1.png)

The original has a white background padding around the QR code. In the processed version you can actually see the padding because the transformation changed the color to a light gray. The processed version's edges are "fuzzy" and the resulting file size is twice as big as the original. 

Using the NearestNeighbor resampler gives a much, much better result:

![image](https://user-images.githubusercontent.com/3506334/137365020-803694a7-2e08-4c1e-9ca4-370c65f67a99.png)

.. but every other image is worse.

Is there a setting I'm missing to make Lanczos5 work well with QR codes (and other images like them)? I've tried a few images that are just text (both blocky and thin) and the process above with Lanczos5 looks really good. If not, is there any recommendation on how to move forward with this? I've thought of a couple ideas:
1. If the result file is larger than original, just return the original as the processed version
2. Detect somehow which resampler to use at runtime - I don't even know where to start with this

**Comment by antonfirsov** [comment, marked as answer]
@natehitze not an expert on resamplers, but the questions I would ask at this point:
- Does the default Bicubic resampler really produce sub-optimal results?
- What about other resamplers in the library?

> Detect somehow which resampler to use at runtime

I'm afraid picking the resampler based on the image contents is already a Machine Learning area, and leads very far.

**Reply by natehitze** [reply]
Here is Bicubic:
![image](https://user-images.githubusercontent.com/3506334/137503803-75fc8855-09f8-49fb-9703-8f6a4146aaf0.png)

I've tried most of the resamplers in `KnownResampler` and all I've tried except NearestNeighbor give more or less the same result.

**Reply by JimBobSquarePants** [reply]
This is all expected behaviour. Resizing images requires sampling and averaging surrounding pixels. Each algorithm has its trade offs and is better for different scenarios.

Nearest neighbour works well for monochrome images like QR codes (in fact it’s the only algorithm that does) because it only takes a single sample with no averaging.

This website contains a treasure trove of information about how resampling works.

http://entropymine.com/imageworsener/

**Reply by natehitze** [reply]
Thanks! I've implemented a simple algorithm to detect if the image is monochrome and use that to determine which resampler to use.
